azkaban-common/src/main/java/azkaban/project/ProjectManager.java

```diff
@@ -86,12 +98,115 @@ public ProjectManager(ProjectLoader loader, Props props) {
+    }
+
+    @SuppressWarnings("unused")
+    public void shutdown() {
+      shutdown = true;
+      this.interrupt();
+    }
+
+    public void run() {
+      while (!shutdown) {
+        synchronized (this) {
+          try {
+            long currentTime = System.currentTimeMillis();
+
+            if (currentTime - CLEANER_THREAD_WAIT_INTERVAL_MS > lastProjectFileCleanTime) {
+              cleanProjectFilesAndRemoveProject();
+              lastProjectFileCleanTime = currentTime;
+            }
+
+            if (currentTime - CLEANER_THREAD_WAIT_INTERVAL_MS > lastProjectCleanTime) {
+              purgeRemovedProjects();
+              lastProjectCleanTime = currentTime;
+            }
+
+            logger.info("Cleaner Thread completed. Waiting!!!!!");
+            wait(CLEANER_THREAD_WAIT_INTERVAL_MS);
+          } catch (ProjectManagerException e) {
+            logger.error(this.getName() + "failed due to: " + e.getMessage());
+          } catch (InterruptedException e) {
+            logger.info("Interrupted. Probably to shut down.");
+          }
+        }
+      }
+    }
+
+    private void cleanProjectFilesAndRemoveProject() throws ProjectManagerException {
+      User deleter = new User(DELETER);
+      long currentTime = System.currentTimeMillis();
+
+      Set<Integer> scheduledProjectIds = projectLoader.fetchProjectIdsByEventType(EventType.SCHEDULE);
+      logger.info("Scheduled Projects count: " + scheduledProjectIds.size());
+
+      List<Project> projects = projectsById.values()
+              .stream()
+              .filter(project -> !scheduledProjectIds.contains(project.getId()) &&
+                      currentTime - project.getLastModifiedTimestamp() > projectFilesRetentionMs)
+              .collect(Collectors.toList());
+
+      logger.info("Project files to clean: " + projects.size());
+
+      for(Project project : projects) {
+        logger.debug("Cleaning project files for project: " + project.getId());
+        projectLoader.cleanProjectFiles(project);
+        logger.debug("Removing project: " + project.getId());
+        removeProject(project, deleter);
+      }
+    }
+
+    private void purgeRemovedProjects() throws ProjectManagerException {
+      long currentTime = System.currentTimeMillis();
+
+      List<Project> inactiveProjects = projectLoader.fetchAllInactiveProjects();
+      logger.info("Inactive projects count: " + inactiveProjects.size());
+
+      Set<Integer> purgedProjectIds = projectLoader.fetchProjectIdsByEventType(EventType.PURGE);
+      logger.info("Purged projects count: " + purgedProjectIds.size());
+
+      List<Project> projectsToPurge = inactiveProjects
+              .stream()
+              .filter(project -> !purgedProjectIds.contains(project.getId()) &&
+                      currentTime - project.getLastModifiedTimestamp() > removedProjectRetentionMs)
+              .collect(Collectors.toList());
+
+      logger.info("Projects to purge count: " + projectsToPurge.size());
```

Comment:
can you club these 3 logs into 1 and that too is projectsToPurge.size() > 0 to avoid spamming logs

Reply:
This was coming only once per day, so made it different for better readability, we can have it in one line also